runtime/lib/Handlers/api/utils.ts

```diff
@@ -43,90 +59,160 @@ export const getVariable = (path: string, data: any) => {
   return stringToNumIfNumeric(curData);
 };
 
-export const ReduceKeyValue = (values: { key: string; val: string }[]) =>
-  values.reduce<Record<string, string>>((acc, { key, val }) => {
-    if (key) {
-      acc[key] = val;
-    }
-    return acc;
-  }, {});
+const validateUrl = async (urlString: string, runtime: Runtime) => {
+  let url;
+  try {
+    url = new URL(urlString);
+  } catch (err) {
+    throw new VError(`url: ${urlString} could not be parsed`, VError.HTTP_STATUS.BAD_REQUEST);
+  }
+
+  const { hostname } = url;
+
+  if (hostname.toLowerCase() === 'localhost') {
+    throw new VError(`url hostname cannot be localhost: ${urlString}`, VError.HTTP_STATUS.BAD_REQUEST);
+  }
+
+  // DNS.resolve returns an array of ips
+  const ips = validator.isIP(hostname)
+    ? [hostname]
+    : await DNS.resolve(hostname).catch(() => {
+        throw new VError(`cannot resolve hostname: ${hostname}`, VError.HTTP_STATUS.BAD_REQUEST);
+      });
+
+  PROHIBITED_IP_RANGES.forEach((prohibitedRange) => {
+    ips.forEach((ip) => {
+      if (ipRangeCheck(ip, prohibitedRange)) {
+        throw new VError(`url resolves to IP: ${ip} in prohibited range`, VError.HTTP_STATUS.BAD_REQUEST);
+      }
+    });
+  });
+
+  if (await runtime.outgoingApiLimiter.addHostnameUseAndShouldThrottle(hostname)) {
+    // if the use of the hostname is high, delay the api call but let it happen
+    await new Promise((resolve) => setTimeout(resolve, THROTTLE_DELAY));
+  }
+};
 
-export const formatRequestConfig = (data: APINodeData) => {
-  const { method, bodyInputType, headers, body, params, url, content } = data;
+// eslint-disable-next-line sonarjs/cognitive-complexity
+export const getResponse = async (data: APINodeData, runtime: Runtime) => {
+  const { method, bodyInputType, headers, body, params, url } = data;
 
-  const options: AxiosRequestConfig = {
+  let content;
+  try {
+    content = JSON.parse(data.content);
+  } catch (e) {
+    ({ content } = data);
+  }
+
+  BLACKLISTED_URLS.forEach((regex) => {
+    if (regex.test(url)) {
+      throw new VError('url endpoint is blacklisted', VError.HTTP_STATUS.BAD_REQUEST);
+    }
+  });
```

Comment:
`Array.prototype.some` could be used here too.

```suggestion
  if (BLACKLISTED_URLS.some((regex) => regex.test(url))) {
    throw new VError('url endpoint is blacklisted', VError.HTTP_STATUS.BAD_REQUEST);
  }
```